lib/web_ui/lib/src/engine/canvaskit/text.dart

```diff
@@ -553,74 +563,138 @@ class CkParagraph extends ManagedSkiaObject<SkParagraph>
-          builder.pop();
-          break;
-        case _ParagraphCommandType.pushStyle:
-          builder.pushStyle(command.style!);
-          break;
-        case _ParagraphCommandType.addPlaceholder:
-          builder._addPlaceholder(command.placeholderStyle!);
-          break;
+    final bool constraintsChanged = _lastLayoutConstraints != constraints;
+    if (didRebuildSkiaObject || constraintsChanged) {
+      _lastLayoutConstraints = constraints;
+      // TODO(het): CanvasKit throws an exception when laid out with
+      // a font that wasn't registered.
+      try {
+        paragraph.layout(constraints.width);
+        _alphabeticBaseline = paragraph.getAlphabeticBaseline();
+        _didExceedMaxLines = paragraph.didExceedMaxLines();
+        _height = paragraph.getHeight();
+        _ideographicBaseline = paragraph.getIdeographicBaseline();
+        _longestLine = paragraph.getLongestLine();
+        _maxIntrinsicWidth = paragraph.getMaxIntrinsicWidth();
+        _minIntrinsicWidth = paragraph.getMinIntrinsicWidth();
+        _width = paragraph.getMaxWidth();
+        _boxesForPlaceholders = skRectsToTextBoxes(paragraph.getRectsForPlaceholders());
+      } catch (e) {
+        printWarning('CanvasKit threw an exception while laying '
+            'out the paragraph. The font was "${_paragraphStyle._fontFamily}". '
+            'Exception:\n$e');
+        rethrow;
       }
     }
 
-    final SkParagraph result = builder._buildCkParagraph();
-    if (_lastLayoutConstraints != null) {
-      // We need to set the Skia object early so layout works.
-      rawSkiaObject = result;
-      this.layout(_lastLayoutConstraints!);
+    return paragraph;
+  }
+
+  // Caches laid out paragraphs and synchronously reclaims memory if there's
+  // memory pressure.
+  //
+  // On May 26, 2021, 500 seemed like a reasonable number to pick for the cache
+  // size. At the time a single laid out SkParagraph used 100KB of memory. So,
+  // 500 items in the cache is roughly 50MB of memory, which is not too high,
+  // while at the same time enough for most use-cases.
+  //
+  // TODO(yjbanov): this strategy is not sufficient for the use-case where a
+  //                lot of paragraphs are laid out _and_ rendered. To support
+  //                this use-case without blowing up memory usage we need this:
+  //                https://github.com/flutter/flutter/issues/81224
+  static SynchronousSkiaObjectCache _paragraphCache = SynchronousSkiaObjectCache(500);
+
+  /// Marks this paragraph as eligible for GC.
+  ///
+  /// Puts this paragraph in a [SynchronousSkiaObjectCache], which will delete it
+  /// if there memory pressure to do so. This protects our memory usage from
+  /// blowing up if within a single frame the framework needs to layout a lot of
+  /// paragraphs. One common use-case is `ListView.builder`, which needs to layout
+  /// more of its content than it actually renders to compute the scroll position.
+  void release() {
```

Comment:
I think this name is a little confusing. `release` makes me think we are telling the cache to delete this paragraph ASAP. I think something like `addToCache` or `markUsed` would make more sense

Reply:
Good point. Done.